core/auth.py

```
@@ -0,0 +1,356 @@
+        if api_key not in self.rate_limits:
+            self.rate_limits[api_key] = []
+
+        # Clean old timestamps (older than 1 hour)
+        cutoff = now - timedelta(hours=1)
+        self.rate_limits[api_key] = [
+            ts for ts in self.rate_limits[api_key] if ts > cutoff
+        ]
+
+        # Check rate limit
+        if len(self.rate_limits[api_key]) >= key.rate_limit:
+            raise ValueError(f"Rate limit exceeded. Maximum {key.rate_limit} requests per hour.")
```

Ruff (0.14.10)

Avoid specifying long messages outside the exception class

(TRY003)